popovers: Restructure the Move messages and Move topic modals.

saubhagya-patel opened the pull request:
This PR refactors the move messages and move topic modals to follow standard patterns from other modals (e.g. the invite user modal). It includes the following changes:

- Refactored the intro text for the move messages and move topic modals.
- New Labels are introduced in the move messages and move topic modals:
  - New channel
  - New topic
  - Which messages should be moved? <- in the move messages modal
     (no changes to checkboxes)
- Some CSS adjustments as mentioned in the [CZO thread](https://chat.zulip.org/#narrow/channel/9-issues/topic/topic.20box.20too.20short.20in.20.22move.20messages.22/near/1962636).

<!-- Describe your pull request here.-->

Fixes: #32168.
[CZO thread](https://chat.zulip.org/#narrow/channel/9-issues/topic/topic.20box.20too.20short.20in.20.22move.20messages.22/near/1962636)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
  <summary>1. Move messages modal</summary>
  
  | Before                               | After                                |
  |--------------------------------------|--------------------------------------|
  | ![Screenshot 2024-11-16 151315](https://github.com/user-attachments/assets/0fed3fbf-e49d-4d6a-a066-eb1b0b65b5ca) | ![Screenshot 2024-11-16 151541](https://github.com/user-attachments/assets/15088447-6745-4555-9e57-1ba9abdb098a) |

</details>


<details>
  <summary>2. Move topic modal</summary>

  | Before                               | After                                |
  |--------------------------------------|--------------------------------------|
  | ![Screenshot 2024-11-16 151218](https://github.com/user-attachments/assets/5df24465-2e81-4301-8e1c-15d6003e2b9c) | ![Screenshot 2024-11-16 151748](https://github.com/user-attachments/assets/3dbacd33-2b4d-4f34-9c64-5675d453e890) |

</details>


<details>
 <summary>3. Rename topic modal <em>(no changes)</em></summary>

  | Before                               | After                                |
  |--------------------------------------|--------------------------------------|
  | ![Screenshot 2024-11-16 151457](https://github.com/user-attachments/assets/249c5b76-b2b3-4e98-8ee0-fbd9a2e6fe81) | ![Screenshot 2024-11-16 151902](https://github.com/user-attachments/assets/f8d696b1-caf8-4e05-9098-6daf216669ee) |

</details>



---
<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on web/templates/move_topic_to_stream.hbs:
```diff
@@ -1,26 +1,50 @@
-        <input name="new_topic_name" type="text" class="move_messages_edit_topic modal_text_input" autocomplete="off" value="{{topic_name}}" {{#if disable_topic_input}}disabled{{/if}} />
+        <div class="input-group">
+            {{#unless only_topic_edit}}
+            <label for="new_topic" class="modal-field-label">New topic</label>
```
`for` should never point to something that doesn't exist.

timabbott replied:
Probably you want to add an `move-topic-new-topic-name` ID on the input and then point this `for` at that.

timabbott commented on web/templates/move_topic_to_stream.hbs:
```diff
@@ -1,26 +1,50 @@
+<p>
+    {{#tr}}
+    You are moving all messages from <z-stream></z-stream>
+    {{#*inline "z-stream"}}<b>{{> inline_decorated_stream_name stream=stream show_colored_icon=true}}</b>{{/inline}}
```
This change seems to remove the `white-space-preserve-wrap`; is that intended? It seems likely that class was there for a reason.

saubhagya-patel replied:
Yes, removing the class `white-space-preserve-wrap` was intentional. This class caused the text to preserve white spaces and wrap. After modifying the intro-text, preserving white spaces resulted in unnecessary spacing. Please refer to the attached screenshots. Thank you.

<details>
<summary>Screenshots</summary>

| Screenshot with the class applied | Screenshot with the class removed |
|-----------------------------------|-----------------------------------|
| ![Screenshot 2024-11-19 091631](https://github.com/user-attachments/assets/159827d2-0511-4590-b269-45d936a59088) | ![Screenshot 2024-11-19 091819](https://github.com/user-attachments/assets/878e3fce-8417-4f12-b05a-9c4daf187f04) |


</details>

timabbott commented on web/templates/move_topic_to_stream.hbs:
```diff
@@ -1,26 +1,50 @@
-        </select>
+        <div class="input-group">
+            <label for="messages_moved">{{t "Which messages should be moved?"}}</label>
+            <select id="messages_moved" class="message_edit_topic_propagate modal_select bootstrap-focus-style">
```
`messages_moved` is too generic a name for an ID, which must be unique across the app. Try `move-topic-select-propagation-mode`.

timabbott commented:
This looks roughly right but note the 3 potential issues above. Once you've fixed those, @sayamsamal can do a next review.

saubhagya-patel commented:
I have updated this PR based on the feedback received from Tim Abbott. The following changes were made:
- Introduced an ID `move-topic-new-topic-name` for the new topic name field.
- Changed the ID of the Message Edit propagate mode select group from `messages_moved` to `move-topic-select-propagation-mode`.
This PR is ready for review. Thank you.

sayamsamal commented on web/styles/popovers.css:
```diff
@@ -917,12 +915,17 @@ ul.popover-group-menu-member-list {
     .move_messages_edit_topic {
-        margin-bottom: 10px;
+        margin-bottom: unset;
+        width: 97%;
```
From the `width: 97%`, that I see over here, seems like setting the `width` to `100%` is causing the input box to overflow? Did you happen to find a reason for that?

sayamsamal replied:
Maybe try setting `box-sizing` to `border-box` on the input element, and then adjusting the padding? I think you'll also need remove/modify the `height` property on the input element as well.

saubhagya-patel replied:
Yes, setting the `width` to `100%` caused the input box to overflow. I tried to identify the cause but was unsure if the behavior was related to the `simplebar` classes. I will keep this in mind and will put in more effort to debug if similar issues arise in the future.

Also, setting `box-sizing` to `border-box` works well. Thank you for the valuable feedback.

saubhagya-patel commented:
I have updated this PR, as per the review received in [this](https://github.com/zulip/zulip/pull/32370#discussion_r1850845534) comment.
Changes made:
 - The input element with class `move_messages_edit_topic`, now has following properties:
   - `box-sizing` as `border-box`
   - `width` is set to `100%`
   - `height` is set to `auto`
 
Since modifying the `height` property of the input element resolved the issue, I didn't adjust the `padding`.
Please let me know if further adjustments are needed. Thank you.

sayamsamal commented on web/templates/move_topic_to_stream.hbs:
```diff
@@ -1,26 +1,50 @@
     <p>{{t "Rename topic to:" }}</p>
     {{else if from_message_actions_popover}}
-    <p>{{t "Move messages to:" }}</p>
+    <p>
+        {{#tr}}
+        You are moving messages from <z-stream></z-stream>
+        {{#*inline "z-stream"}}<b>{{> inline_decorated_stream_name stream=stream show_colored_icon=true}}</b>{{/inline}}
+        {{/tr}}
+        <i class="fa fa-angle-right" aria-hidden="true"></i>
+        {{#tr}}<b>{topic_name}.</b>{{/tr}}
+    </p>
```
I think this block belongs with the one above it.

sayamsamal replied:
Maybe we could structure this template in general. The `only_topic_edit` conditional looks like it is being used multiple times – being used when only topic renaming is possible. We could consolidate the elements to show for `only_topic_edit`.

sayamsamal commented:
The previous set of changes LGTM @saubhagya-patel, thanks! I have left another feedback regarding the structure of the handlebar template, which I believe will improve the whole readability of this template. You can put the structuring changes in a separate commit, so it doesn't overlap with the main motive of this issue, which is to standardize the move/rename modal.

saubhagya-patel commented:
I have updated this PR as per the review received in [this](https://github.com/zulip/zulip/pull/32370#issuecomment-2490801359) comment.

The following changes were made:
- The first commit contains no changes and addresses [this](https://github.com/zulip/zulip/issues/32168) issue.
- The second commit includes the following updates:
   - Refactored the `move_topic_to_stream.hbs` template displayed during the process of moving messages or renaming/moving a topic to improve its structure and readability.
   - Set the `width` of the `<p>` tag in the template to `fit-content`.
     
     <details>
     <summary>Details of the added <code>width</code>.</summary>
     
     I added this because I found that the intro-text at the top of the modal in some cases caused some layout bugs. Kindly refer to the screenshot.

     | No width style applied | Width style applied |
     |------------------------|---------------------|
     | ![Screenshot 2024-11-22 003505](https://github.com/user-attachments/assets/2f282bca-9182-4a8d-a575-ad54c368de12) | ![Screenshot 2024-11-22 003343](https://github.com/user-attachments/assets/d17194cd-f008-4349-b7d6-edb8e64c44d5) |

     </details>
   
   - Removed the unused styles for `.fa-angle-right` in the template.

     <details>
     <summary>Details of the unused styles.</summary>
     These styles were previously applied to the angle-right(>) sign between the channel and topic in the modal. Since the modal no longer uses the sign, I decided to remove them. Kindly refer to the screenshot.

     ![Screenshot 2024-11-22 002711](https://github.com/user-attachments/assets/b0dfe6d8-5d51-4924-a594-727e190abce1)

     </details>
     

Please let me know if any further adjustments are required. Thank you.

sayamsamal commented on web/styles/popovers.css:
```diff
@@ -917,14 +915,25 @@ ul.popover-group-menu-member-list {
+        flex-flow: column wrap;
         justify-content: flex-start;
 
+        p {
+            width: fit-content;
+        }
+
```
What is the purpose of this change? Also, we want to avoid using tag names directly like this, they are too general and are prone to overflow onto other elements unintentionally. Please define a class to apply this property if needed.

saubhagya-patel replied:
I apologize for using the tag names directly and will remember not to use them directly in the future.

The style here was applied because in some cases the intro-text of the modal caused layout bugs specifically additional space between the intro-text. I am attaching some screenshots. However, in some cases, it remained fine.

<details>
  <summary>Screenshots of the added <code>width</code> on the modal.</summary>
  
  | No width style applied | Width style applied |
  |------------------------|---------------------|
  | ![Screenshot 2024-11-22 003505](https://github.com/user-attachments/assets/2f282bca-9182-4a8d-a575-ad54c368de12) | ![Screenshot 2024-11-22 003343](https://github.com/user-attachments/assets/d17194cd-f008-4349-b7d6-edb8e64c44d5) |
  | ![Screenshot 2024-11-25 232059](https://github.com/user-attachments/assets/c362c5de-5238-4f3e-9490-c85a2ead330a) | ![Screenshot 2024-11-25 232135](https://github.com/user-attachments/assets/ca2a0ae4-ec95-4cca-804d-18f8d0bd25db) |
  
</details>

Please note that in the first row of the attached screenshots there is some extra space below the intro-text and in the second row it isn't. I found this while testing on different stream and topic names. Hence, I found that applying the `width` style would help us eliminate these layout bugs. 
Thank you.

sayamsamal replied:
Interesting, and the setting the width does seem to fix it.

saubhagya-patel commented:
I have updated this PR based on the review received in [this](https://github.com/zulip/zulip/pull/32370#discussion_r1856386265) comment.

Changes made:
- The `width` style applied on the intro-text of the move topic modal is now applied using a class named `move-topic-modal-intro-text`

Please let me know if any further adjustments are required. Thank you.

saubhagya-patel commented:
I have updated this PR and made the following changes:

- Resolved conflicts.

- The third commit adds a default value ("change_all") to the `propagate_mode` field in the `stream_popover.ts` file:
  - The enum schema for the `propagate_mode` field was introduced in the commit [93185dd](https://github.com/zulip/zulip/commit/93185dd7a68b0944b3ab071c3dd1d7653cea4cd7#diff-9c69b59bb892d719f83b2152182c1ec7e86f37c581aaa7c06b275876d38cf0fe). When moving a topic (via the move topic popover), the function `get_params_from_form()` was returning an `undefined` value, leading to an error. Adding the default value resolves this issue. I have attached screenshots of the before and after, kindly check them.

Screenshots of moving topic:
  - <details>
    <summary>Without default value</summary>
    
     ![move topic error](https://github.com/user-attachments/assets/c47d5030-12b4-4636-a397-922a078cd0b9)
    
  </details>
  
  - <details>
    <summary>With default value</summary>
    
    ![move topic no error](https://github.com/user-attachments/assets/b34af45c-be1e-4d59-936c-edec5a374e27)
    
  </details>

Please let me know if any further changes are required. Thank you.

saubhagya-patel commented:
In the latest push the following changes are included:
- Resolved conflicts.
- The third commit is undone as it is no longer needed, the `propagate_mode` is now treated as optional (commit - [`d2c4659`](https://github.com/zulip/zulip/commit/d2c465960d0f9f23daa60d9bb845e20fd8042d99)).

Please let me know if any further changes are required. Thank you.

sayamsamal commented:
Thanks @saubhagya-patel, good catch on the issue with the `propagate_mode`, and good to know that it has been fixed as well. Passing it ahead to @timabbott for integration.

alya commented:
Playing with the PR, let's drop the intro line, and put the source into the heading instead. So the heading would be:

- Move messages from #channel > topic
- Move topic from #channel > topic

If the channel > topic name doesn't fit on one line, we can abbreviate it.

alya commented:
I think we can make the "Rename topic" modal be consistent with the others. So the title would be:

> Rename topic #channel > topic

And the field label can be "New topic".

saubhagya-patel commented:
Following the review received, I have updated this PR.  
I have added a third commit, which includes the following changes:  

- As mentioned in [this comment](https://github.com/zulip/zulip/pull/32370#issuecomment-2518499147), the intro text is removed from the "Move Topic" and "Move Message(s)" modals. Instead, it is placed in the modal heading. The heading is abbreviated using ellipsis (`...`) where it doesn't fit on one line.  

  <details>
    <summary>Screenshot of Move Topic modal.</summary>
    
    ![Screenshot 2024-12-06 005421](https://github.com/user-attachments/assets/138b0827-1ba0-4964-8934-81e658f5d2eb)  
    
  </details>  

  <details>
    <summary>Screenshot of Move Messages modal.</summary>
    
    ![Screenshot 2024-12-06 005300](https://github.com/user-attachments/assets/62c731bb-fea0-4e75-83f1-0dc71e5be677) 
     
  </details>  

  <details>
    <summary>Screenshot of Move Topic modal with ellipsis.</summary>
    
    ![Screenshot 2024-12-06 005507](https://github.com/user-attachments/assets/78acacce-aa56-4d84-8b83-da756466cd8e)  
    
  </details>  

  <details>
    <summary>Screenshot of Move Messages modal with ellipsis.</summary>
    
    ![Screenshot 2024-12-06 005530](https://github.com/user-attachments/assets/c6ca0631-eb94-4fa3-bd1a-8a1afa1ec39d)  
    
  </details>  

- As mentioned in [this comment](https://github.com/zulip/zulip/pull/32370#issuecomment-2518504190), the "Rename Topic" modal also follows the above layout changes. The field label "New Topic" is also included.  

  <details>
    <summary>Screenshot of Rename Topic modal.</summary>
    
    ![Screenshot 2024-12-06 010010](https://github.com/user-attachments/assets/b2a0e947-55fe-4d03-9b19-6d24432ec440)  
    
  </details>  

  <details>
    <summary>Screenshot of Rename Topic modal with ellipsis.</summary>
    
    ![Screenshot 2024-12-06 005915](https://github.com/user-attachments/assets/bc12b471-fb24-4af8-aa6e-b58465847eb7)  
    
  </details>

@alya, please note that I have included the changes in a separate commit for testing purposes. I can edit the commits once the changes look good. Please let me know if any further changes are required. Thank you.

alya commented:
We need a bit of space after the `#`-- you can see how there's space between it an the channel name in other parts of the app.

alya commented:
Otherwise, it looks good. You should squash any extra commits to prepare your PR for code review.

saubhagya-patel commented:
I have updated this PR based on the review received. I have made the following changes:
- As mentioned in [this](https://github.com/zulip/zulip/pull/32370#issuecomment-2521649133) comment to add some space after the `#` icon, a `margin-right` of `5px` has been applied to the `#` icon. I have added this, taking into account the `flex-gap` between the `#` icon and the stream name, just like the message-header.

  <details>
    <summary>Screenshot of Move messages modal with updated margin.</summary>
    
    ![Screenshot 2024-12-07 231631](https://github.com/user-attachments/assets/12a7391d-a0fa-4d9c-9b95-e2fe071f6bfb)
    
  </details>

  <details>
    <summary>Screenshot of the referenced message-header.</summary>
    
    ![Screenshot 2024-12-07 231343](https://github.com/user-attachments/assets/9ae7f1e9-7563-4646-b9f1-582ece23b547)
  </details>
  
- Also, as mentioned in [this](https://github.com/zulip/zulip/pull/32370#issuecomment-2521649447) comment:
   > Otherwise, it looks good. You should squash any extra commits to prepare your PR for code review.

   I will definitely keep this in mind in the future. To address this, I have rewritten the commit history.

@alya, this PR is ready for review. I hope the changes look good to you. Please let me know if any further adjustments are needed. Thank you.

alya commented:
Thanks! That looks better, but we still want to have more space to the left of the `#` than to the right, so that it's easy to parse visually as being associated with the channel name.

alya commented:
Please post screenshots with all three privacy icons once you've had a chance to tweak it.

saubhagya-patel commented:
I have updated this PR based on the review received. I have made the following changes:

- As stated in [this](https://github.com/zulip/zulip/pull/32370#issuecomment-2532953455) comment, I have applied `padding-left: 5px;` and `padding-right: 3px;` to the `#` icon to ensure proper space around it. The `padding-left` is greater than the `padding-right` to make sure that the space to the left of the `#` is larger than the space to the right.
   
   <details>
   <summary>Screenshot of Move messages modal with updated spacing (with the hashtag icon).</summary>
   
   ![Screenshot 2024-12-11 153506](https://github.com/user-attachments/assets/c9ccfa6e-b92a-49b1-abac-81b8897c71b4)

   </details>
   
   <details>
   <summary>Screenshot of Move messages modal with updated spacing (with the globe icon).</summary>
   
   ![Screenshot 2024-12-11 153549](https://github.com/user-attachments/assets/729e12dc-a29d-4758-8e8e-0929ed4f009a)

   </details>
   
   <details>
   <summary>Screenshot of Move messages modal with updated spacing (with the lock icon).</summary>
   
   ![Screenshot 2024-12-11 153625](https://github.com/user-attachments/assets/46593469-bc9d-4e62-b4e5-488afb7c7f1c)

   </details>
   
@alya, this PR is ready for review. I hope the updated changes look good to you. Please let me know if any further adjustments are needed. Thank you.

timabbott commented:
I think the space between "from" and the channel icon is too small. (Also note some merge conflicts from #32155 just merging).

timabbott commented on web/src/stream_popover.ts:
```diff
@@ -345,6 +345,7 @@ export async function build_move_topic_to_stream_popover(
     message?: Message,
```
"Refactor" is not a correct word to use in this commit message; it is only correct if there are no functional changes.

"restructure" would be a fine word to use.

saubhagya-patel replied:
Sure. Thank you for the feedback. I will surely remember this in the future.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes have been made:

- The merge conflicts have been resolved.
- The commit message has been updated to use "restructure" in place of "refactor," as stated in [this](https://github.com/zulip/zulip/pull/32370#discussion_r1880922183) comment.
- The space around the `#` icon in the "Move messages" and "Move topic" modals has been updated following [this](https://github.com/zulip/zulip/pull/32370#issuecomment-2537082181) comment. Now, the `#` icon has a `padding-right` of `5px` and a `padding-left` of `10px`. I agree that the updated changes indeed look better.

   <details>
   <summary>Screenshot of "Move messages" modal with updated spacing (with the hashtag icon).</summary>
   
   ![Screenshot 2024-12-12 184154](https://github.com/user-attachments/assets/9c415be1-d16d-42ee-aa6f-6ce101b7827a)

   </details>
   
   <details>
   <summary>Screenshot of "Move messages" modal with updated spacing (with the globe icon).</summary>
   
   ![Screenshot 2024-12-12 184305](https://github.com/user-attachments/assets/703cb11c-9663-4eb5-9ddb-42ea9b3368fb)

   </details>
   
   <details>
   <summary>Screenshot of "Move messages" modal with updated spacing (with the lock icon).</summary>
   
   ![Screenshot 2024-12-12 184721](https://github.com/user-attachments/assets/56836b2a-4239-4482-bbc6-004a219d1fc6)

   </details>
   
- Added the attribute `name="propagate_mode"` in the "Which messages should be moved?" select group, which was removed in PR [#32155](https://github.com/zulip/zulip/pull/32155). Removing this caused an assertion error.
  
   <details>
   <summary>Screen recording of error caused.</summary>
   
   ![propagate_mode error](https://github.com/user-attachments/assets/6c7f9626-a3f6-4773-b68d-6766c394f0ba)
   
   </details>
   
   <details>
   <summary>Screenshot of the error caused.</summary>
   
   ![Screenshot 2024-12-12 185246](https://github.com/user-attachments/assets/4cdb16d3-1072-4449-af55-752c87d7e97c)

   </details>
   
@timabbott, this PR is now ready for review. I hope the changes look good to you. Please let me know if any further adjustments are needed. Thank you.

timabbott commented:
Looks good. Merged after changing that `10px` offset to the left of the channel type indicator to 3px. 

Next time, please do bug fixes in `main` as their own commit and report the bug in #issues when you see it -- since those may be things we need to fix in a timely fashion. Seemed OK in this case.

timabbott commented:
Oops, cb75feafbb4af934544e99e6c1619140953259af was in a follow-up push.

saubhagya-patel commented:
Thanks @timabbott for the insight. I will remember this in the future.
